### PR TITLE
Only mutate the required contexts when changing branch protections

### PR DESCRIPTION
**Pull request by stevekuznetsov**
The configuration for the mungers today only allows for the required
contexts to be explicitly stated for a protected branch. Until we expose
the other types of protections for a branch, our branch protection
requests should make sure not to mutate those fields. Furthermore, now
that we know there are types of protections we do not control, it is not
possible to know that we should remove a branch protection entirely, so
that functionality is removed. If all types of protections are exposed
through the configuration in the future, that functionality can return.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/kind bug
/area mungegithub

/cc @fejta @bgrant0607 @rmmh @cjwagner

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Working on some unit tests, but review is welcome

/hold

File: mungegithub/github/github.go [hunk 1]
```diff
@@ -738,7 +709,31 @@ func (config *Config) SetBranchProtection(name string, contexts []string) error
...
+	return config.setBranchProtection(name, request)
+}
+
+// unchangedRestrictionRequest generates a request that will
+// not make any changes to the teams and users that can merge
+// into a branch
+func unchangedRestrictionRequest(restrictions *github.BranchRestrictions) *github.BranchRestrictionsRequest {
+	request := &github.BranchRestrictionsRequest{}
+
+	for _, user := range restrictions.Users {
```
Comment (rmmh) [inline mungegithub/github/github.go]: `Teams and users restrictions are only available for organization-owned repositories.`

If this is nullable, wrap in an if-guard?

**Comment by stevekuznetsov** [comment]
Hmm alternatively I think we can use [this endpoint](https://developer.github.com/v3/repos/branches/#update-required-status-checks-of-protected-branch) if we want to bump the API client lib

**Comment by cjwagner** [comment]
> Hmm alternatively I think we can use this endpoint if we want to bump the API client lib

That may not be sufficient. For branches that don't already have branch protection do we need to set `EnforceAdmins` to false or is that the default?

**Comment by stevekuznetsov** [comment]
Good point @cjwagner ... in any case even the latest `master` of `go-github` doesn't yet support that endpoint so I think for now we should use this approach, then we can go upstream to add support for other endpoints as we feel it is necessary.

**Comment by stevekuznetsov** [comment]
@rmmh added `nil`-guards

File: mungegithub/github/github.go [hunk 2]
```diff
@@ -738,7 +709,42 @@ func (config *Config) SetBranchProtection(name string, contexts []string) error
...
+// not make any changes to the teams and users that can merge
+// into a branch
+func unchangedRestrictionRequest(restrictions *github.BranchRestrictions) *github.BranchRestrictionsRequest {
+	request := &github.BranchRestrictionsRequest{
+		Users: []string{},
+		Teams: []string{},
+	}
+
+	if restrictions == nil {
+		return request
```
Comment (cjwagner) [inline mungegithub/github/github.go]: Should nil be returned here instead of an empty request?

Comment (stevekuznetsov) [reply mungegithub/github/github.go]: Should be same functionality, but yes we can return `nil`

**Comment by stevekuznetsov** [comment]
We won't catch new fields with this refactor anyway, there is a _lot_ of work required to mock out for unit tests and it's unclear how useful it will be in enforcing the invariant we want (current state is not mutated).

/hold cancel

**Comment by fejta** [comment]
/assign @cjwagner @rmmh 
/uncc @fejta
LGTM generally

**Comment by bgrant0607** [comment]
Thanks to all who helped resolve this quickly!

**Pull request merged**